addons/website_slides/static/src/js/slides_course_join.js

```diff
@@ -39,25 +73,35 @@ var CourseJoinWidget = publicWidget.Widget.extend({
             }
         }).popover('show');
     },
+    /**
+     * @private
+     */
+    _signInAndJoinCourse: function () {
+        var url = this._createLoginRedirectUrl();
+        this.trigger_up('before_join_course_redirect', {url: url});
+    },
 
     //--------------------------------------------------------------------------
-    // Handlers
+    // Public
     //--------------------------------------------------------------------------
-
     /**
-     * @private
+     * @public
+     * @param {integer} channelId
      */
-    _onClickJoin: function (event) {
-        var channelId = this.channelId || $(event.currentTarget).data('channel-id');
+    joinChannel: function (channelId) {
         var self = this;
         this._rpc({
             route: '/slides/channel/join',
             params: {
                 channel_id: channelId,
             },
         }).then(function (data) {
-            if (! data.error) {
-                location.reload();
+            if (!data.error) {
+                if (self.isQuiz) {
+                    self.trigger_up('submit_after_join');
```

Comment:
Again, just like "beforeJoin", there should be an "afterJoin".
But it does not have to be a promise, just a regular function that acts as a callback.

That way you can get rid of the "isQuiz" variable and make the join widget agnostic to its parent "type".